### DISCUSSION TITLE
How to write custom callback with monitor

### DISCUSSION THREAD

**Discussion by ZENGYIMING-EAMON** [Lightning-AI/pytorch-lightning#13045] [Lightning Trainer API: Trainer, LightningModule, LightningDataModule]
I am using PL-1.6.1.

I am using the official `pl.callbacks.ModelCheckpoint` with `monitor: 'some_lss/dataloader_idx_1', mode: 'min'` and it works fine.

Now I write a custom callback `class CustomCallback(pl.Callback):` to save some images at the end of train epoch, 
by  writing `def on_train_epoch_end(self,trainer,pl_module):` under `class CustomCallback(pl.Callback):`.

Now I save it every 10 epochs using `if trainer.current_epoch%50==0:` under `def on_train_epoch_end(self,trainer,pl_module):`.

But I want to change it to save every time the model meets a lower loss, i.e., every time the model meets a min `'some_lss/dataloader_idx_1'`. How can I incorporate the `monitor: 'some_lss/dataloader_idx_1', mode: 'min'` into my custom callback?

**Comment by rohitgr7** [comment]
you can access this `some_lss/dataloader_idx_1` using `trainer.callback_metrics['some_lss/dataloader_idx_1']` and use `on_validation_end` hook to write your logic to checkpoint if the new value is lower than the current lowest.

```py
class CustomCallback(pl.Callback):

    def __init__(self, mode='min', monitor='some_lss/dataloader_idx_1'):
        ...
        torch_inf = torch.tensor(np.Inf)
        mode_dict = {"min": torch_inf, "max": -torch_inf}
        self.monitor_op = {"min": torch.lt, "max": torch.gt}[mode]
        self.best_metric_value = mode_dict[mode]

    def on_validation_end(self, trainer, pl_module):
        current_metric = trainer.callback_metrics[self.monitor]
        if self.monitor_op(current_metric, self.best_metric_value):
            self.best_metric_value = current_metric
            # do whatever you want
```
